Two improvements to elpaca-menu-item

Stebalien opened the pull request:
(I'm happy to break this up into separate PRs if desired)

- The first patch improves integration into newer completion frameworks by defining groups and aligning descriptions.
- The second patch switches from let-binding `completion-extra-properties` to using `completion-table-with-metadata`. The former breaks recursive completion.

More complete descriptions can be found in the commit messages.

Stebalien commented on elpaca.el:
```diff
@@ -6,7 +6,7 @@
 ;; Created: Jan 1, 2022
 ;; Keywords: tools, convenience, lisp
-;; Package-Requires: ((emacs "27.1"))
+;; Package-Requires: ((emacs "28.1"))
```
For `string-search`. I can also do it with a regex.

Stebalien replied:
Done.

progfolio commented on elpaca.el:
```diff
@@ -289,12 +289,32 @@ If ID is nil, prompt for item. If INTERACTIVE is non-nil, copy to `kill-ring'."
+                   (props `((group-function
+                             . ,(lambda (cand transform)
+                                  (if-let* ((idx (string-search " " cand)))
+                                      (if transform (substring cand 0 idx) (substring cand (1+ idx)))
```
Probably a micro-optimization, but `substring-no-properties` may be more performant.

Stebalien replied:
done

progfolio commented on elpaca.el:
```diff
@@ -289,12 +289,32 @@ If ID is nil, prompt for item. If INTERACTIVE is non-nil, copy to `kill-ring'."
+                            ;; annotation-function so we can align all currently displayed completions.
+                            (affixation-function
+                             . ,(lambda (completions)
+                                  ;; Increase alignment if necessary, never decrease.
+                                  (setq align (apply #'max align
+                                                     (mapcar (lambda (cand) (+ 2 (string-width cand))) completions)))
```
Can you explain why alignment would never need to be decreased?
Also, why do we add 2 to the width of each candidate? Spacing?

Stebalien replied:
Done. I've moved `2` out into a separate constant. I'm happy to make it a customizable variable (possibly a customizable string?). Another option is to use `marginalia-separator` if bound, falling back on two spaces otherwise.

progfolio commented on elpaca.el:
```diff
@@ -289,12 +289,32 @@ If ID is nil, prompt for item. If INTERACTIVE is non-nil, copy to `kill-ring'."
+                                   (lambda (cand) (list
+                                              cand nil
+                                              (concat (propertize " " 'display `(space :align-to ,align))
+                                                      (propertize
+                                                       (plist-get (cdr (alist-get cand candidates nil nil #'equal))
+                                                                  :description)
```
If a candidate does not have a `:description`, `propertize` will be called with a `nil` first argument, which will signal a wrong-type-argument error.

Stebalien replied:
fixed.

progfolio reviewed (changes requested):
Thanks for this!
It's definitely a better completion experience.
I left a couple comments, questions.
Another thing to consider:

Is it worth breaking out the lambdas in `props` into named functions so they can be debugged easier? The reason I ask is because I ran into this error when testing in a minimal environment:

```emacs-lisp
Debugger entered--Lisp error: (wrong-type-argument stringp nil)
  string-width(nil)
  #<subr F616e6f6e796d6f75732d6c616d626461_anonymous_lambda_59>(("0blayout MELPA" nil #(" Layout grouping with ease" 0 1 (display (space :align-to 48)) 1 26 (face completions-annotations))))
  completion--insert-strings((("0blayout MELPA" nil #(" Layout grouping with ease" 0 1 (display ...) 1 26 (face completions-annotations))) ("0x0 MELPA" nil #(" Upload sharing to 0x0.st" 0 1 (display ...) 1 25 (face completions-annotations))) ("0xc MELPA" nil #(" Base conversion made easy" 0 1 (display ...) 1 26 (face completions-annotations))) ("2bit MELPA" nil #(" Library for reading data from 2bit files" 0 1 (display ...) 1 41 (face completions-annotations))) ("750words MELPA" nil #(" Emacs integration and Org exporter for 750words.com" 0 1 (display ...) 1 52 (face completions-annotations))) ("@ MELPA" nil #(" Multiple-inheritance prototype-based objects DSL" 0 1 (display ...) 1 49 (face completions-annotations))) ("a MELPA" nil #(" Associative data structure functions" 0 1 (display ...) 1 37 (face completions-annotations))) ("aa-edit-mode MELPA" nil #(" Major mode for editing AA(S_JIS Art) and .mlt file" 0 1 (display ...) 1 51 (face completions-annotations))) ("aangit MELPA" nil #(" Quickly scaffold new Angular apps with Aangit" 0 1 (display ...) 1 46 (face completions-annotations))) ("aas MELPA" nil #(" Snippet expansions mid-typing" 0 1 (display ...) 1 30 (face completions-annotations))) ("abc-mode MELPA" nil #(" Major mode for editing abc music files" 0 1 (display ...) 1 39 (face completions-annotations))) ("abgaben MELPA" nil #(" Review and correct assignments received by mail" 0 1 (display ...) 1 48 (face completions-annotations))) ("abl-mode MELPA" nil #(" Python TDD minor mode" 0 1 (display ...) 1 22 (face completions-annotations))) ("abridge-diff MELPA" nil #(" Abridge long line-based diff hunks, including in magit" 0 1 (display ...) 1 55 (face completions-annotations))) ("abs-mode MELPA" nil #(" Major mode for the modeling language Abs" 0 1 (display ...) 1 41 (face completions-annotations))) ("abyss-theme MELPA" nil #(" A dark theme with contrasting colours" 0 1 (display ...) 1 38 (face completions-annotations))) ("ac-alchemist MELPA" nil #(" Auto-complete source for alchemist" 0 1 (display ...) 1 35 (face completions-annotations))) ("ac-c-headers MELPA" nil #(" Auto-complete source for C headers" 0 1 (display ...) 1 35 (face completions-annotations))) ("ac-capf MELPA" nil #(" Auto-complete source with completion-at-point" 0 1 (display ...) 1 46 (face completions-annotations))) ("ac-clang MELPA" nil #(" Auto Completion source by libclang for GNU Emacs" 0 1 (display ...) 1 49 (face completions-annotations))) ("ac-dcd MELPA" nil #(" Auto Completion source for dcd for GNU Emacs" 0 1 (display ...) 1 45 (face completions-annotations))) ("ac-emacs-eclim MELPA" nil #(" Auto-complete source for eclim" 0 1 (display ...) 1 31 (face completions-annotations))) ("ac-emmet MELPA" nil #(" Auto-complete sources for emmet-mode's snippets" 0 1 (display ...) 1 48 (face completions-annotations))) ("ac-emoji MELPA" nil #(" Auto-complete source of Emoji" 0 1 (display ...) 1 30 (face completions-annotations))) ("ac-etags MELPA" nil #(" Etags/ctags completion source for auto-complete" 0 1 (display ...) 1 48 (face completions-annotations))) ("ac-geiser MELPA" nil #(" Auto-complete backend for geiser" 0 1 (display ...) 1 33 (face completions-annotations))) ("ac-haskell-process MELPA" nil #(" Haskell auto-complete source which uses the current haskell process" 0 1 (display ...) 1 68 (face completions-annotations))) ("ac-helm MELPA" nil #(" Helm interface for auto-complete" 0 1 (display ...) 1 33 (face completions-annotations))) ("ac-html MELPA" nil #(" Auto complete source for html tags and attributes" 0 1 (display ...) 1 50 (face completions-annotations))) ("ac-html-angular MELPA" nil #(" Auto complete angular15 data for `ac-html' and `company-web'" 0 1 (display ...) 1 61 (face completions-annotations))) ("ac-html-bootstrap MELPA" nil #(" Auto complete bootstrap3/fontawesome classes for `ac-html' and `company-web'" 0 1 (display ...) 1 77 (face completions-annotations))) ("ac-html-csswatcher MELPA" nil #(" Css/less class/id completion with `ac-html' or `company-web'" 0 1 (display ...) 1 61 (face completions-annotations))) ...) #f(lambda (cand transform) [t] (let* ((idx ...)) (if idx (if transform ... ...) cand))))
```

The environment was created via:

```emacs-lisp
(elpaca-test
  :interactive t
  :ref local ;; with your pull request checked out, built
  :init
  (setf (alist-get 'emacs package--builtin-versions nil 'remove) nil))
```

And then trying to complete via `elpaca-info`.

Stebalien commented on elpaca.el:
```diff
@@ -289,12 +289,32 @@ If ID is nil, prompt for item. If INTERACTIVE is non-nil, copy to `kill-ring'."
-                            (concat " " (plist-get (cdr (alist-get s candidates nil nil #'equal))
-                                                   :description)))))
-                   (choice (completing-read (or elpaca-overriding-prompt "Menu Item: ") candidates nil t)))
+            (let* ((align 20)
```
Yes. I've moved it out into a separate constant (happy to make it a custom option and/or variable).

Stebalien commented:
I think I've addressed all the feedback. Unfortunately, I can't get rid of the lambda for the affixation function unless we want to let alignment decrease. But in my testing, that caused quite a bit of re-alignment jitter where keeping the max alignment feels more natural.

progfolio commented:
> I think I've addressed all the feedback. Unfortunately, I can't get rid of the
> lambda for the affixation function unless we want to let alignment decrease. But
> in my testing, that caused quite a bit of re-alignment jitter where keeping the
> max alignment feels more natural.

Thank you.
Unfortunately, I was wrong about substring-no-properties.
It looks like the group name loses its properties if we go with that approach.

I pushed a refactor which introduces an option for the alignment.
I also did not include a minimum spacing. Everything appears to work on my end without it.
Can you confirm on your end?

Stebalien commented on elpaca.el:
```diff
@@ -276,6 +276,29 @@ Values for each key are that of the right-most plist containing that key."
+      (substring candidate (if transform 0 (1+ space)) (when transform space))
+    candidate))
+
+(defcustom elpaca-min-affixation-column 20
```
I'd consider calling this something like `elpaca-min-description-column` (or `-annotation-`). The `affixation-function` both prepends and appends (although we're only using it for the latter) and it's mostly an implementation detail.

progfolio replied:
Good point. I've changed it to that.

Stebalien commented on elpaca.el:
```diff
@@ -276,6 +276,29 @@ Values for each key are that of the right-most plist containing that key."
+ITEMS is an alist mapping completion candidates to menu ITEMS."
+  (lambda (completions)
+    (let ((column (max (or elpaca-min-affixation-column 0)
+                       (or (cl-loop for c in completions maximize (string-width c)) 0))))
```
Ooh. I never knew `cl-loop` had a maximize verb.

progfolio replied:
cl-loop has a lot going for it. `sum` and `minimize` are also handy.

Stebalien commented on elpaca.el:
```diff
@@ -276,6 +276,29 @@ Values for each key are that of the right-most plist containing that key."
+(defun elpaca--completion-group-by-menu (candidate transform)
+  "Return CANDIDATE menu if TRANSFORM non-nil, otherwise CANDIDATE item."
+  (if-let* ((space (string-match-p " " candidate)))
+      (substring candidate (if transform 0 (1+ space)) (when transform space))
```
This is technically shorter, but I have to say it's harder to read.

progfolio replied:
I'm not overly concerned in this case since it's essentially a single line.
If I happen across it in the future and have any trouble reading it, we can change it for clarity.

Stebalien commented:
>  I also did not include a minimum spacing. Everything appears to work on my end without it.

This is because of the group transformation (the group is taken into account when computing the width). If you disable your current completion framework, enable `icomplete-vertical-mode`, and set `completion-groups` to `nil`; you'll see sources squished up against descriptions.

---

In terms of the refactor, my only nit is that this will now let the alignment decrease. It's not a huge issue, but it differs from the behavior of marginalia.

progfolio commented:
> This is because of the group transformation (the group is taken into account when computing the width). If you disable your current completion framework, enable `icomplete-vertical-mode`, and set `completion-groups` to `nil`; you'll see sources squished up against descriptions.

Thanks for pointing this out. I think a fair compromise is prepending a single space to each description.

> In terms of the refactor, my only nit is that this will now let the alignment decrease. It's not a huge issue, but it differs from the behavior of marginalia.

I see what you mean after experimenting more. I've added back the previous method, which will only grow.
let me know what you think of latest version.

Stebalien commented:
LGTM!

progfolio commented:
Merged. Thanks again!

Stebalien commented:
No, thank you for the quick review! Also, Elpaca is awesome.
